Conan path: find_package(GTest) 在 CI 失败 — Configure 通过但 build 报 gtest.h 找不到

## 背景

PR #9 第一次让 Conan 路径上 CI（`linux-gcc-conan` job，`gcc-relwithdebinfo-conan` preset），暴露出长期存在但从未被验证的问题。该 job 已在 `.github/workflows/ci.yml` 暂时注释掉，等本 issue 解决后取消注释。

## 症状

- `conan install . --output-folder=build/gcc-relwithdebinfo-conan --build=missing -s compiler.cppstd=23` 成功
- `cmake --preset gcc-relwithdebinfo-conan` configure 成功（toolchain 加载、summary 正常打印 "Build tests: ON"）
- `cmake --build --preset gcc-relwithdebinfo-conan` build 失败：

```
modules/01_basics/tests/test_basics.cpp:8:10: fatal error: gtest/gtest.h: No such file or directory
    8 | #include <gtest/gtest.h>
```

## 矛盾点

`cmake/Dependencies.cmake` 的逻辑：

```cmake
find_package(GTest CONFIG QUIET)
if(NOT GTest_FOUND)
    find_package(GTest MODULE QUIET)
endif()
if(NOT GTest_FOUND)
    message(FATAL_ERROR "GoogleTest not found...")
endif()
```

Configure 没 FATAL_ERROR → `GTest_FOUND` 一定是 TRUE。但 build 时 `GTest::gtest_main` target 的 `INTERFACE_INCLUDE_DIRECTORIES` 没生效。说明 `find_package` 在某种意义上"成功"了，却没拿到正确的 include 路径。

## 已确认的 CI 日志细节

```
conanfile.txt: CMakeToolchain generated: conan_toolchain.cmake
conanfile.txt: CMakeToolchain generated: <repo>/build/gcc-relwithdebinfo-conan/CMakePresets.json
conanfile.txt: CMakeToolchain generated: <repo>/CMakeUserPresets.json
```

CMakeDeps 的输出**没有**出现在日志里（可能是 silent generation），所以不知道 `GTestConfig.cmake` 实际写到了哪。

`conan_toolchain.cmake` 默认会 `list(PREPEND CMAKE_PREFIX_PATH \"\${CMAKE_CURRENT_LIST_DIR}\")`，所以 CMAKE_PREFIX_PATH 的设置不是问题源。

## 根因候选（按概率排）

### 1. Module 模式兜底匹配上半残的 system gtest

去掉 `cmake_layout` 后 CONFIG 模式可能因路径细节没找到 `GTestConfig.cmake`，回退到 MODULE 模式。CMake 自带 `FindGTest.cmake` 搜系统路径——某些 ubuntu-24.04 runner 镜像装了 `libgtest-dev` 的 lib 但缺头，导致 \"找到库但 INCLUDE_DIRS 空\"，target 链上后 build 期才暴露。

### 2. CMakeDeps 输出位置与 `find_package` 搜索路径不对齐

去掉 `[layout] cmake_layout` 后 CMakeDeps 的 `generators_folder` 行为有 Conan 2.x 版本差异：
- 有的版本：写在 `<output-folder>/`
- 有的版本：仍写在 `<output-folder>/build/<config>/generators/`（不受 `--output-folder` 控制）
- 有的版本：写在某个包子目录下

`conan_toolchain.cmake` 的 `CMAKE_PREFIX_PATH` 只 prepend `<output-folder>/`，找不到深层子目录的 `GTestConfig.cmake`。

### 3. Conan 生成的 target 名与代码不匹配

`mcpp_add_test` 用 `GTest::gtest_main`（大写）。gtest 的 Conan recipe 应该已设 `cmake_target_name = \"GTest::gtest\"`，但旧版 recipe 可能仍导出小写 `gtest::gtest`，链接阶段就该 fail 才对——除非被某种 silent skip 跳过。

## 下一步：诊断步骤

在 CI 重新加 job 时，`Conan install` 之后插：

```yaml
- name: Debug — list generator output
  if: matrix.uses-conan
  run: |
    echo \"=== build/<preset>/ tree ===\"
    ls -laR build/${{ matrix.preset }}/ | head -100
    echo \"=== Files matching *Config.cmake ===\"
    find build/${{ matrix.preset }}/ -name '*Config.cmake' -o -name '*-config.cmake'
    echo \"=== conan_toolchain.cmake content (first 80 lines) ===\"
    head -80 build/${{ matrix.preset }}/conan_toolchain.cmake
```

5 分钟就能定位是 1 / 2 / 3 哪个。

## 修复路径候选

### A. 显式指定 cmake_find_mode

`conanfile.txt` 加：

```ini
[conf]
tools.cmake.cmakedeps:new=will_break_next
```

或在 conanfile.py 形式里设 `cmake_find_mode = \"config\"`。强制 CMakeDeps 只生成 config 模式文件，避免 Module 兜底兜到坏的 system gtest。

### B. 恢复 cmake_layout，改用 per-build-type preset 路径

恢复 `[layout] cmake_layout` 后，路径变成 `<output-folder>/build/<BuildType>/generators/conan_toolchain.cmake`，需要每个 leaf preset 单独指定路径：

```json
{
    \"name\": \"gcc-debug-conan\",
    \"cacheVariables\": {
        \"CMAKE_TOOLCHAIN_FILE\": \"\${sourceDir}/build/gcc-debug-conan/build/Debug/generators/conan_toolchain.cmake\"
    }
}
```

代价：21 个 `*-conan` leaf preset 的 cache 变量都要展开成自己的完整路径，无法靠继承的 `_conan` 父 preset 收敛。

### C. 放弃自定义 `*-conan` preset，改用 Conan 自动生成的 CMakeUserPresets.json

`conan install` 后 Conan 已经在仓库根自动写出 `CMakeUserPresets.json`，里面有 `conan-release` / `conan-debug` 等 preset，路径与 toolchain 完全自洽。代价：README 表格里那 21 个 `*-conan` preset 名变成"文档展示用"，实际命令是 `cmake --preset conan-<build_type>`。

### D. 显式设置 `CMAKE_FIND_PACKAGE_PREFER_CONFIG=ON`

Dependencies.cmake 在 `find_package(GTest)` 前加 `set(CMAKE_FIND_PACKAGE_PREFER_CONFIG ON)`，强制走 CONFIG 模式，禁止 Module 兜底。如果根因 1 是真的，这一步立刻见效。

## 验收标准

- 取消注释 `linux-gcc-conan` matrix 项
- CI 该 job 通过
- 在 docs/conan-guide.md 补一段诊断笔记说明最终采用了哪条路径

## 参考

- PR #9 三个相关 commit：`fix(conan): _conan preset 的 toolchain 路径漏了 generators/ 子目录`、`fix(conan): 移除 cmake_layout`、`ci: 暂时注释掉 linux-gcc-conan job`
- CI 失败 run：https://github.com/with-fair-wind/ModernCpp/actions/runs/24982204282/job/73146833436

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -41,19 +41,17 @@ jobs:
             cxx: g++-13
             extra-config: -DMCPP_ENABLE_SANITIZERS=ON
 
-          # TODO(conan): re-enable once CMakeDeps + find_package(GTest CONFIG)
-          # path is debugged. Symptom is "fatal error: gtest/gtest.h: No such
-          # file or directory" at build time even though configure succeeds.
-          # Likely root cause: <Pkg>Config.cmake placement under --output-folder
-          # doesn't match where CMAKE_PREFIX_PATH (set by conan_toolchain.cmake)
-          # tells find_package to look. Steps and conanfile.txt notes are kept
-          # below in this file so re-enabling is just removing the comment-out.
-          # - name: linux-gcc-conan
-          #   os: ubuntu-24.04
-          #   preset: gcc-relwithdebinfo-conan
-          #   cc: gcc-13
-          #   cxx: g++-13
-          #   uses-conan: true
+          # Conan path. Must pass `-s build_type=` matching the preset's
+          # CMAKE_BUILD_TYPE — otherwise CMakeDeps generates per-config target
+          # files for the wrong config and INTERFACE_INCLUDE_DIRECTORIES is
+          # empty at build time. cmake/Dependencies.cmake catches this.
+          - name: linux-gcc-conan
+            os: ubuntu-24.04
+            preset: gcc-relwithdebinfo-conan
+            cc: gcc-13
+            cxx: g++-13
+            uses-conan: true
+            build-type: RelWithDebInfo
 
           - name: windows-msvc
             os: windows-2022
@@ -131,6 +129,7 @@ jobs:
           conan install . \
             --output-folder=build/${{ matrix.preset }} \
             --build=missing \
+            -s build_type=${{ matrix.build-type }} \
             -s compiler.cppstd=23
 
       # ------------------------------------------------------------------

--- a/cmake/Dependencies.cmake
+++ b/cmake/Dependencies.cmake
@@ -38,5 +38,29 @@ if(MCPP_BUILD_TESTS)
             "To skip tests entirely pass -DMCPP_BUILD_TESTS=OFF.")
     endif()
 
+    # Defensive check for the Conan + CMakeDeps build_type mismatch trap:
+    # CMakeDeps only emits per-config target files (GTest-Target-<lower>.cmake)
+    # for the build_type passed to `conan install -s build_type=...`. If that
+    # doesn't match CMAKE_BUILD_TYPE, the IMPORTED targets get created (so
+    # GTest_FOUND is TRUE) but their INTERFACE_INCLUDE_DIRECTORIES is wrapped
+    # in $<$<CONFIG:OtherType>:...> generator expressions that evaluate to
+    # empty under our build_type — configure passes, build later explodes with
+    # `gtest/gtest.h: No such file`. Catch this at configure time.
+    if(TARGET GTest::gtest_main AND CMAKE_BUILD_TYPE)
+        get_target_property(_mcpp_gtest_inc GTest::gtest_main INTERFACE_INCLUDE_DIRECTORIES)
+        if(_mcpp_gtest_inc)
+            string(FIND "${_mcpp_gtest_inc}" "$<$<CONFIG:" _mcpp_has_genex)
+            string(FIND "${_mcpp_gtest_inc}" "$<$<CONFIG:${CMAKE_BUILD_TYPE}>" _mcpp_has_match)
+            if(NOT _mcpp_has_genex EQUAL -1 AND _mcpp_has_match EQUAL -1)
+                message(FATAL_ERROR
+                    "GoogleTest was found but has no include directories for "
+                    "CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}.\n"
+                    "This usually means `conan install` was run with a different "
+                    "build_type than the CMake preset. Re-run with: "
+                    "-s build_type=${CMAKE_BUILD_TYPE}")
+            endif()
+        endif()
+    endif()
+
     include(GoogleTest)
 endif()

--- a/conanfile.txt
+++ b/conanfile.txt
@@ -11,3 +11,10 @@ CMakeToolchain
 # Without a layout, --output-folder=X puts conan_toolchain.cmake directly at
 # X/conan_toolchain.cmake, which the _conan preset then references as
 # ${sourceDir}/build/${presetName}/conan_toolchain.cmake. Simple and stable.
+#
+# IMPORTANT: pass `-s build_type=<Type>` matching the CMake preset's build type
+# (e.g. -s build_type=RelWithDebInfo for *-relwithdebinfo-conan presets).
+# CMakeDeps only emits per-config target files for the requested build_type;
+# a mismatch leaves INTERFACE_INCLUDE_DIRECTORIES empty under generator
+# expressions, configure passes, build fails with `gtest/gtest.h: No such file`.
+# cmake/Dependencies.cmake guards against this with a configure-time check.

--- a/docs/conan-guide.md
+++ b/docs/conan-guide.md
@@ -415,6 +415,17 @@ build"，是日常开发的安全默认。
 
 覆盖 profile 里的 `build_type`。本仓库需要这一行，因为 profile 默认 Release。
 
+> ⚠️ **必须与 preset 的 `CMAKE_BUILD_TYPE` 完全一致**。`gcc-debug-conan` 用
+> `-s build_type=Debug`，`gcc-relwithdebinfo-conan` 用 `-s build_type=RelWithDebInfo`，
+> 类推。CMakeDeps 只为请求的 build_type 生成 per-config 目标文件
+> （`GTest-Target-<lower>.cmake` + `GTest-<lower>-x86_64-data.cmake`）；mismatch
+> 时 `find_package(GTest)` 仍然返回成功（`GTest::gtest_main` target 在
+> `GTestTargets.cmake` 里被无条件 `add_library(... INTERFACE IMPORTED)`），但
+> `INTERFACE_INCLUDE_DIRECTORIES` 被 `$<$<CONFIG:OtherType>:...>` 包起来，对当前
+> build_type 求值为空 —— configure 通过、build 时炸 `gtest/gtest.h: No such file`。
+> `cmake/Dependencies.cmake` 已经加了一道 configure 期检查捕获这种情况，撞上时会
+> 直接 FATAL_ERROR 提示重跑命令。
+
 #### `-s compiler.cppstd=23`
 
 覆盖 C++ 标准。本仓库要 C++23，profile 默认可能是 17。

